### PR TITLE
Fix deprecated statement

### LINKED ISSUE #3304: Replace `torch.cuda.amp.autocast` with `torch.amp.autocast('cuda',...)`
Fix the warning:
```
 FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
```
by replacing everywhere in the codebase:
```diff
- torch.cuda.amp.autocast(...
+ torch.amp.autocast("cuda", ...
```

### PR DESCRIPTION
**Pull request by Gulin7**
Fixes #3304

Description:
Replaced all occurances of torch.cuda.amp.autocast with torch.amp.autocast('cuda',...)

### CODE REVIEW

File: examples/notebooks/CycleGAN_with_torch_cuda_amp.ipynb [hunk 1]
```diff
@@ -896,7 +896,8 @@
     "id": "vrJls4p-FRcA"
    },
    "source": [
-    "from torch.cuda.amp import autocast, GradScaler\n",
+    "from torch.cuda.amp import GradScaler\n",
+    "from torch.amp import autocast\n",
```
Comment (vfdev-5) [inline examples/notebooks/CycleGAN_with_torch_cuda_amp.ipynb:900]:
```suggestion
    "from torch.amp import autocast\n",
    "from torch.cuda.amp import GradScaler\n",
```

File: examples/references/segmentation/pascal_voc2012/main.py [hunk 2]
```diff
@@ -6,7 +6,8 @@
 import torch
 
 try:
-    from torch.cuda.amp import autocast, GradScaler
+    from torch.amp import autocast
+    from torch.cuda.amp import GradScaler
 except ImportError:
     raise RuntimeError("Please, use recent PyTorch version, e.g. >=1.6.0")
```
Comment (vfdev-5) [inline examples/references/segmentation/pascal_voc2012/main.py:12]: `torch.amp` was added in 1.12.0, let's update the error message.

File: examples/transformers/main.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 import torch.nn as nn
 import torch.optim as optim
 import utils
-from torch.cuda.amp import autocast, GradScaler
+from torch.amp import autocast, GradScaler
```
Comment (vfdev-5) [inline examples/transformers/main.py:10]: Let's import GradScaler as previously:
```python
from torch.cuda.amp import GradScaler
```

File: ignite/engine/__init__.py [hunk 4]
```diff
@@ -185,7 +185,7 @@ def supervised_training_step_amp(
     """
 
     try:
-        from torch.cuda.amp import autocast
+        from torch.amp import autocast
     except ImportError:
         raise ImportError("Please install torch>=1.6.0 to use amp_mode='amp'.")
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:190]: Same, here update min pytorch version to 1.12.0

File: ignite/engine/__init__.py [hunk 5]
```diff
@@ -726,15 +726,15 @@ def supervised_evaluation_step_amp(
         Added `model_fn` to customize model's application on the sample
     """
     try:
-        from torch.cuda.amp import autocast
+        from torch.amp import autocast
     except ImportError:
         raise ImportError("Please install torch>=1.6.0 to use amp_mode='amp'.")
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:731]: Same, here update min pytorch version to 1.12.0

File: tests/ignite/engine/test_create_supervised.py [hunk 6]
```diff
@@ -631,7 +631,7 @@ def test_create_supervised_evaluator():
 
     # older versions didn't have the autocast method so we skip the test for older builds
     if Version(torch.__version__) >= Version("1.6.0"):
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:633]: Same, here update min pytorch version to 1.12.0

File: tests/ignite/engine/test_create_supervised.py [hunk 7]
```diff
@@ -641,7 +641,7 @@ def test_create_supervised_evaluator_on_cpu():
...
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:643]: Same, here update min pytorch version to 1.12.0

File: tests/ignite/engine/test_create_supervised.py [hunk 8]
```diff
@@ -652,7 +652,7 @@ def test_create_supervised_evaluator_traced_on_cpu():
...
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:654]: Same, here update min pytorch version to 1.12.0

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @Gulin7
I left few comments to address

**Comment by Gulin7** [comment]
@vfdev-5 I addressed your comments, tag me and let me know if there is anything else that needs improvement.

File: ignite/metrics/kl_divergence.py [hunk 9]
```diff
@@ -9,7 +9,7 @@
 
 __all__ = ["KLDivergence"]
 
-TORCH_VERSION_GE_160 = Version(torch.__version__) >= Version("1.6.0")
+TORCH_VERSION_GE_160 = Version(torch.__version__) >= Version("1.12.0")
```
Comment (vfdev-5) [inline ignite/metrics/kl_divergence.py:12]: Let's revert this

File: ignite/metrics/js_divergence.py [hunk 10]
```diff
@@ -8,7 +8,7 @@
 
 __all__ = ["JSDivergence"]
 
-TORCH_VERSION_GE_160 = Version(torch.__version__) >= Version("1.6.0")
+TORCH_VERSION_GE_160 = Version(torch.__version__) >= Version("1.12.0")
```
Comment (vfdev-5) [inline ignite/metrics/js_divergence.py:11]: Same here, unrelated change to revert

**Comment by vfdev-5** [comment]
Some tests are failing as well: https://github.com/pytorch/ignite/actions/runs/12045408777/job/33647387903

File: examples/notebooks/CycleGAN_with_torch_cuda_amp.ipynb [hunk 11]
```diff
@@ -1,31 +1,4 @@
 {
- "nbformat": 4,
```
Comment (vfdev-5) [inline examples/notebooks/CycleGAN_with_torch_cuda_amp.ipynb:2]: Can you please make just necessary changes without all unrelated updates?

File: ignite/engine/__init__.py [hunk 12]
```diff
@@ -412,7 +412,7 @@ def _check_arg(
             try:
                 from torch.cuda.amp import GradScaler
             except ImportError:
-                raise ImportError("Please install torch>=1.6.0 to use scaler argument.")
+                raise ImportError("Please install torch>=1.12.0 to use scaler argument.")
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:415]: GradScaler was since 1.6.0, so let's keep the previous message

File: tests/ignite/engine/test_create_supervised.py [hunk 13]
```diff
@@ -447,21 +447,21 @@ def test_create_supervised_trainer_apex_error():
 def mock_torch_cuda_amp_module():
     with patch.dict(
         "sys.modules",
-        {"torch.cuda.amp": None, "torch.cuda.amp.grad_scaler": None, "torch.cuda.amp.autocast_mode": None},
+        {"torch.cuda.amp": None, "torch.cuda.amp.grad_scaler": None, "torch.amp.autocast_mode": None},
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:450]:
```suggestion
        {"torch.amp": None, "torch.cuda.amp.grad_scaler": None, "torch.amp.autocast_mode": None},
```

File: tests/ignite/engine/test_create_supervised.py [hunk 14]
```diff
@@ -447,21 +447,21 @@ def test_create_supervised_trainer_apex_error():
...
-    with pytest.raises(ImportError, match="Please install torch>=1.6.0 to use amp_mode='amp'."):
+    with pytest.raises(ImportError, match="Please install torch>=1.12.0 to use amp_mode='amp'."):
         _test_create_supervised_trainer(amp_mode="amp")
-    with pytest.raises(ImportError, match="Please install torch>=1.6.0 to use scaler argument."):
+    with pytest.raises(ImportError, match="Please install torch>=1.12.0 to use scaler argument."):
         _test_create_supervised_trainer(amp_mode="amp", scaler=True)
 
 
-@pytest.mark.skipif(Version(torch.__version__) < Version("1.5.0"), reason="Skip if < 1.5.0")
+@pytest.mark.skipif(Version(torch.__version__) < Version("1.12.0"), reason="Skip if < 1.5.0")
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:464]:
```suggestion
@pytest.mark.skipif(Version(torch.__version__) < Version("1.12.0"), reason="Skip if < 1.12.0")
```

**Comment by vfdev-5** [comment]
@Gulin7 there is a failing test with error message:
```

=========================== short test summary info ============================
FAILED tests/ignite/engine/test_create_supervised.py::test_create_supervised_trainer_amp_error - AssertionError: Regex pattern did not match.
 Regex: 'Please install torch>=1.12.0 to use scaler argument.'
 Input: "Please install torch>=1.12.0 to use amp_mode='amp'."
======================== 1 failed, 3 warnings in 2.71s =========================
```
and also there is this comment to address: https://github.com/pytorch/ignite/pull/3307#discussion_r1861935669

Can you please fix them?

File: tests/ignite/engine/test_create_supervised.py [hunk 15]
```diff
@@ -457,7 +457,7 @@ def test_create_supervised_trainer_amp_error(mock_torch_cuda_amp_module):
         _test_create_supervised_trainer_wrong_accumulation(trainer_device="cpu", amp_mode="amp")
     with pytest.raises(ImportError, match="Please install torch>=1.12.0 to use amp_mode='amp'."):
         _test_create_supervised_trainer(amp_mode="amp")
-    with pytest.raises(ImportError, match="Please install torch>=1.12.0 to use scaler argument."):
+    with pytest.raises(ImportError, match="Please install torch>=1.12.0 to use amp_mode='amp'."):
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:460]: This does not looks like a correct change

Comment (Gulin7) [reply tests/ignite/engine/test_create_supervised.py:460]: Where in the test should "Please install torch>=1.12.0 to use scaler argument." be raised?

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py:460]: The original message is raised in the code

File: ignite/engine/__init__.py [hunk 16]
```diff
@@ -461,7 +461,7 @@ def create_supervised_trainer(
         amp_mode: can be ``amp`` or ``apex``, model and optimizer will be casted to float16 using
             `torch.cuda.amp <https://pytorch.org/docs/stable/amp.html>`_ for ``amp`` and
             using `apex <https://nvidia.github.io/apex>`_ for ``apex``. (default: None)
-        scaler: GradScaler instance for gradient scaling if `torch>=1.6.0`
+        scaler: GradScaler instance for gradient scaling if `torch>=1.12.0`
```
Comment (vfdev-5) [inline ignite/engine/__init__.py:464]:
```suggestion
        scaler: GradScaler instance for gradient scaling if `torch>=1.6.0`
```

**Review by vfdev-5 — APPROVED** [review]
LGTM
